### PR TITLE
Avoid using typing-imported symbols for runtime edits

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

When we go to import a symbol (in order to use it for an autofix), we first check whether the symbol is available in the current scope. This PR ensures that when we use an existing symbol, we first verify that it was created in the appropriate context. That is, we shouldn't reuse `defaultdict` here:

```py
import typing

if typing.TYPE_CHECKING:
    from collections import defaultdict


def f(x: typing.DefaultDict[str, str]) -> None:
    ...
```

Since `typing.DefaultDict` needs to be available at runtime, as per Python's somewhat tricky rules of annotation availability. However, we _should_ reuse `defaultdict` here:

```py
import typing

if typing.TYPE_CHECKING:
    from collections import defaultdict


def f(x: "typing.DefaultDict[str, str]") -> None:
    ...
```

There's actually a more general issue here, which is that our binding tracking generally ignores preconditions for bound symbols. So, e.g., we would incorrectly reuse `defaultdict` here:

```py
import typing

if 1 > 2:
    from collections import defaultdict


def f(x: typing.DefaultDict[str, str]) -> None:
    ...
```

But, I'm just ignoring that for now.

## Test Plan

Added some additional fixtures for these specific cases.

### CODE REVIEW

File: crates/ruff/src/importer/mod.rs [hunk 1]
```diff
@@ -220,3 +229,13 @@ impl<'a> Importer<'a> {
...
+
+enum Resolution {
+    /// The symbol is available for use.
+    Success(Edit, String),
+    /// The symbol is imported, but the import came after the current location.
+    LateBinding,
+    /// The symbol is imported, but in an incompatible context (e.g., in typing-only context, while
+    /// we're in a runtime context).
+    IncompatibleContext,
+}
```
Comment (charliermarsh) [inline crates/ruff/src/importer/mod.rs:241]: This is perhaps a strange pattern, but I was finding all the `Option<Result<...>>` stuff to be really messy.

File: crates/ruff/src/importer/mod.rs [hunk 2]
```diff
@@ -70,11 +70,16 @@ impl<'a> Importer<'a> {
...
+        match self.get_symbol(module, member, at, semantic_model) {
+            None => self.import_symbol(module, member, at, semantic_model),
+            Some(Resolution::Success(edit, binding)) => Ok((edit, binding)),
+            Some(Resolution::LateBinding) => {
+                bail!("Unable to use existing symbol due to late binding")
+            }
+            Some(Resolution::IncompatibleContext) => {
+                bail!("Unable to use existing symbol due to incompatible context")
+            }
+        }
```
Comment (MichaReiser) [inline crates/ruff/src/importer/mod.rs:82]: Nit: Would it make sense to define our own `Error` type here to allow callers to match on the reason or is this something that we don't need?

File: crates/ruff/src/importer/mod.rs (hunk 1 above)
Comment (MichaReiser) [reply crates/ruff/src/importer/mod.rs:241]: I like this a lot. It communicates the outcome much clearer. We could even go one step further and flatten the Option too, but I think that would make the usage more cumbersome.

**Review by MichaReiser — APPROVED** [review]

**Review by konstin — APPROVED** [review]

File: crates/ruff/src/importer/mod.rs (hunk 2 above)
Comment (charliermarsh) [reply crates/ruff/src/importer/mod.rs:82]: I'm open to it -- do you have an example e.g. from Rome or elsewhere that I could use to learn how to implement that pattern?

Comment (MichaReiser) [reply crates/ruff/src/importer/mod.rs:82]: We have one in ruff :) 

https://github.com/charliermarsh/ruff/blob/4f9d944556910f834985b9fc75f968dc6fca2557/crates/ruff_formatter/src/diagnostics.rs#L1-L162

It's normally:

* Define an `enum`
* Implement `Error` and `Display`
* Maybe: Implement `From<OtherError>` if you want automatic error conversation when using the try operator (the try operator does `ReturnErrorType::from(error_from_the_tried_operation)`

Comment (konstin) [reply crates/ruff/src/importer/mod.rs:82]: `thiserror` can also generate a bunch of the boilerplate in many cases

Comment (charliermarsh) [reply crates/ruff/src/importer/mod.rs:82]: I'm going to do this in a separate PR because I want to cover the `import_symbol` cases too.

**Pull request merged**